svgnative/ports/cairo/CairoSVGRenderer.h

```diff
@@ -0,0 +1,109 @@
+#define SVGViewer_CairoSVGRenderer_h
+
+#include <list>
+#include "SVGRenderer.h"
+#include "cairo.h"
+
+namespace SVGNative
+{
+
+  // SkiaSVGPath object is able to be amended, but Cairo has no API to append something
+  // to existing cairo_path_t object. Thus, we hold cairo_t object. Its surface could
+  // be retrieved in later, by cairo_get_target().
+
+class CairoSVGPath final : public Path
+{
+public:
+    CairoSVGPath();
+    ~CairoSVGPath();
+
+    void Rect(float x, float y, float width, float height) override;
+    void RoundedRect(float x, float y, float width, float height, float cornerRadius) override;
+    void Ellipse(float cx, float cy, float rx, float ry) override;
+
+    void MoveTo(float x, float y) override;
+    void LineTo(float x, float y) override;
+    void CurveTo(float x1, float y1, float x2, float y2, float x3, float y3) override;
+    void CurveToV(float x2, float y2, float x3, float y3) override;
+    void ClosePath() override;
+
+    cairo_t* mPath; // FIXME: rename it to mCairo in future
+
+private:
+    float mCurrentX{};
+    float mCurrentY{};
+};
+
+class CairoSVGTransform final : public Transform
+{
+public:
+    CairoSVGTransform(float a, float b, float c, float d, float tx, float ty);
+
+    void Set(float a, float b, float c, float d, float tx, float ty) override;
+    void Rotate(float r) override;
+    void Translate(float tx, float ty) override;
+    void Scale(float sx, float sy) override;
+    void Concat(const Transform& other) override;
+
+    cairo_matrix_t mMatrix;
+};
+
+class CairoSVGImageData final : public ImageData
+{
+public:
+    CairoSVGImageData(const std::string& base64, ImageEncoding encoding);
+    ~CairoSVGImageData();
+
+    float Width() const override;
+
+    float Height() const override;
+
+
+    cairo_surface_t* mImageData;
+};
+
+class CairoSVGRenderer final : public SVGRenderer
+{
+public:
+    CairoSVGRenderer();
+    ~CairoSVGRenderer();
+
+    std::unique_ptr<ImageData> CreateImageData(const std::string& base64, ImageEncoding encoding) override { return std::unique_ptr<CairoSVGImageData>(new CairoSVGImageData(base64, encoding)); }
+
+    std::unique_ptr<Path> CreatePath() override { return std::unique_ptr<CairoSVGPath>(new CairoSVGPath); }
+
+    std::unique_ptr<Transform> CreateTransform(
+        float a = 1.0, float b = 0.0, float c = 0.0, float d = 1.0, float tx = 0.0, float ty = 0.0) override
+    {
+        return std::unique_ptr<CairoSVGTransform>(new CairoSVGTransform(a, b, c, d, tx, ty));
+    }
+
+    void Save(const GraphicStyle& graphicStyle) override;
+    void Restore() override;
+
+    void DrawPath(const Path& path, const GraphicStyle& graphicStyle, const FillStyle& fillStyle, const StrokeStyle& strokeStyle) override;
+    void DrawImage(const ImageData& image, const GraphicStyle& graphicStyle, const Rect& clipArea, const Rect& fillArea) override;
+
+    void SetCairo(cairo_t* cairo);
+
+private:
+    std::list<double> alphas;
```

Comment:
std::vector might be more efficient for double

Reply:
But it shouldn't be needed with opacity group implementation from cairo graphics. See above.

Reply:
OK, I would remove alphas.